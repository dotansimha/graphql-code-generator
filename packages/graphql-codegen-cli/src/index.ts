export { executeCodegen } from './codegen';
export { generate } from './generate-and-save';
export * from './config';
export * from './init';
export * from './utils/cli-error';
export * from './cli';
export * from './graphql-config';
