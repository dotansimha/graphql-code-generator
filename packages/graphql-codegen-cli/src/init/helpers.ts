import chalk from 'chalk';
import { resolve, relative } from 'path';
import { writeFileSync, readFileSync } from 'fs';
import { Types } from '@graphql-codegen/plugin-helpers';
import detectIndent from 'detect-indent';
import { Answers } from './types';

// Parses config and writes it to a file
export async function writeConfig(answers: Answers, config: Types.Config) {
  const YAML = await import('json-to-pretty-yaml').then(m => ('default' in m ? m.default : m));
  const ext = answers.config.toLocaleLowerCase().endsWith('.json') ? 'json' : 'yml';
  const content = ext === 'json' ? JSON.stringify(config) : YAML.stringify(config);
  const fullPath = resolve(process.cwd(), answers.config);
  const relativePath = relative(process.cwd(), answers.config);

  writeFileSync(fullPath, content, {
    encoding: 'utf-8',
  });

  return {
    relativePath,
    fullPath,
  };
}

// Updates package.json (script and plugins as dependencies)
export async function writePackage(answers: Answers, configLocation: string) {
  // script
  const pkgPath = resolve(process.cwd(), 'package.json');
  const pkgContent = readFileSync(pkgPath, {
    encoding: 'utf-8',
  });
  const pkg = JSON.parse(pkgContent);
  const { indent } = detectIndent(pkgContent);

  if (!pkg.scripts) {
    pkg.scripts = {};
  }

  pkg.scripts[answers.script] = `graphql-codegen --config ${configLocation}`;

  // plugin
  if (!pkg.devDependencies) {
    pkg.devDependencies = {};
  }

  // read codegen's version
  let version: string;

  const dynamicImport = (m: string) => import(m).then(m => ('default' in m ? m.default : m));
  try {
    // Works in tests
    const packageJson = await dynamicImport('../../package.json');
    version = packageJson.version;
  } catch (e) {
    // Works in production (package dist is flat, everything is in the same folder)
    const packageJson = await dynamicImport('./package.json');
    version = packageJson.version;
  }

  answers.plugins.forEach(plugin => {
    pkg.devDependencies[plugin.package] = version;
  });

  if (answers.introspection) {
    pkg.devDependencies['@graphql-codegen/introspection'] = version;
  }

  // If cli haven't installed yet
  pkg.devDependencies['@graphql-codegen/cli'] = version;

  writeFileSync(pkgPath, JSON.stringify(pkg, null, indent));
}

export function bold(str: string) {
  return chalk.bold(str);
}

export function grey(str: string) {
  return chalk.grey(str);
}

export function italic(str: string) {
  return chalk.italic(str);
}
