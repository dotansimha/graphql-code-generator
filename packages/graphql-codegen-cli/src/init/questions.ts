import * as inquirer from 'inquirer';
import { grey } from './helpers';
import { Tags, Answers } from './types';
import { plugins } from './plugins';

export function getQuestions(possibleTargets: Record<Tags, boolean>): inquirer.Question[] {
  return [
    {
      type: 'checkbox',
      name: 'targets',
      message: `What type of application are you building?`,
      choices: getApplicationTypeChoices(possibleTargets)
    },
    {
      type: 'input',
      name: 'schema',
      message: 'Where is your schema?:',
      suffix: grey(' (path or url)'),
      default: 'http://localhost:4000', // matches Apollo Server's default
      validate: (str: string) => str.length > 0
    },
    {
      type: 'input',
      name: 'documents',
      message: 'Where are your operations and fragments?:',
      when: answers => {
        // flatten targets
        // I can't find an API in Inquirer that would do that
        answers.targets = normalizeTargets(answers.targets);

        return answers.targets.includes(Tags.browser);
      },
      default: '**/*.graphql',
      validate: (str: string) => str.length > 0
    },
    {
      type: 'checkbox',
      name: 'plugins',
      message: 'Pick plugins:',
      choices: getPluginChoices,
      validate: (plugins: any[]) => plugins.length > 0
    },
    {
      type: 'input',
      name: 'output',
      message: 'Where to write the output:',
      default: 'src/generated/graphql.ts',
      validate: (str: string) => str.length > 0
    },
    {
      type: 'confirm',
      name: 'introspection',
      message: 'Do you want to generate an introspection file?'
    },
    {
      type: 'input',
      name: 'config',
      message: 'How to name the config file?',
      default: 'codegen.yml',
      validate: (str: string) => {
        const isNotEmpty = str.length > 0;
        const hasCorrectExtension = ['json', 'yml', 'yaml'].some(ext => str.toLocaleLowerCase().endsWith(`.${ext}`));

        return isNotEmpty && hasCorrectExtension;
      }
    },
    {
      type: 'input',
      name: 'script',
      message: 'What script in package.json should run the codegen?',
      validate: (str: string) => str.length > 0
    }
  ];
}

export function getApplicationTypeChoices(possibleTargets: Record<Tags, boolean>) {
  function withFlowOrTypescript(tags: Tags[]) {
    if (possibleTargets.TypeScript) {
      tags.push(Tags.typescript);
    } else if (possibleTargets.Flow) {
      tags.push(Tags.flow);
    } else {
      tags.push(Tags.flow, Tags.typescript);
    }

    return tags;
  }

  return [
    {
      name: 'Backend - API or server',
      key: 'backend',
      value: withFlowOrTypescript([Tags.node]),
      checked: possibleTargets.Node
    },
    {
      name: 'Application built with Angular',
      key: 'angular',
      value: [Tags.angular, Tags.browser, Tags.typescript],
      checked: possibleTargets.Angular
    },
    {
      name: 'Application built with React',
      key: 'react',
      value: withFlowOrTypescript([Tags.react, Tags.browser]),
      checked: possibleTargets.React
    },
    {
      name: 'Application built with Stencil',
      key: 'stencil',
      value: [Tags.stencil, Tags.browser, Tags.typescript],
      checked: possibleTargets.Stencil
    },
    {
      name: 'Application built with other framework or vanilla JS',
      key: 'client',
      value: [Tags.browser, Tags.typescript, Tags.flow],
      checked: possibleTargets.Browser && !possibleTargets.Angular && !possibleTargets.React && !possibleTargets.Stencil
    }
  ];
}

export function getPluginChoices(answers: Answers) {
  return plugins
    .filter(p => p.available(answers.targets))
    .map<inquirer.ChoiceType>(p => {
      return {
        name: p.name,
        value: p,
        checked: p.shouldBeSelected(answers.targets)
      };
    });
}

function normalizeTargets(targets: Tags[] | Tags[][]): Tags[] {
  return [].concat(...targets);
}
