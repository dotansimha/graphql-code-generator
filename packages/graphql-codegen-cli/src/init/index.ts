import inquirer from 'inquirer';
import { Types } from '@graphql-codegen/plugin-helpers';
import { getQuestions } from './questions';
import { guessTargets } from './targets';
import { Answers, Tags } from './types';
import { writeConfig, writePackage, bold } from './helpers';

function log(...msgs: string[]) {
  // eslint-disable-next-line no-console
  console.log(...msgs);
}

export async function init() {
  log(`
    Welcome to ${bold('GraphQL Code Generator')}!
    Answer few questions and we will setup everything for you.
  `);

  const possibleTargets = await guessTargets();

  const answers = await inquirer.prompt<Answers>(getQuestions(possibleTargets));

  // define config
  const config: Types.Config = {
    overwrite: true,
    schema: answers.schema,
    documents: answers.targets.includes(Tags.browser) ? answers.documents : null,
    generates: {
      [answers.output]: {
        plugins: answers.plugins.map(p => p.value),
      },
    },
  };

  // introspection
  if (answers.introspection) {
    addIntrospection(config);
  }

  // config file
  const { relativePath } = await writeConfig(answers, config);

  log(`Fetching latest versions of selected plugins...`);

  // write package.json
  await writePackage(answers, relativePath);

  // Emit status to the terminal
  log(`
    Config file generated at ${bold(relativePath)}
    
      ${bold('$ npm install')}

    To install the plugins.

      ${bold(`$ npm run ${answers.script}`)}

    To run GraphQL Code Generator.
  `);
}

// adds an introspection to `generates`
function addIntrospection(config: Types.Config) {
  config.generates['./graphql.schema.json'] = {
    plugins: ['introspection'],
  };
}
