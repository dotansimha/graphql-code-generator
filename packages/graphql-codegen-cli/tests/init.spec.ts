import bddStdin from 'bdd-stdin';
import { resolve } from 'path';
import { init } from '../src/init';
import { Tags } from '../src/init/types';
import { guessTargets } from '../src/init/targets';
import { plugins } from '../src/init/plugins';
import { bold } from '../src/init/helpers';
import { getApplicationTypeChoices, getPluginChoices } from '../src/init/questions';
import { safeLoad } from 'js-yaml';

jest.mock('fs');
const { version } = require('../package.json');

const SELECT = ' '; // checkbox
const ENTER = '\n';
// const DOWN = bddStdin.keys.down;

const packageJson = {
  withAngular: JSON.stringify({
    version,
    dependencies: {
      '@angular/core': 'x.x.x',
    },
  }),
  withTypescript: JSON.stringify({
    version,
    devDependencies: {
      typescript: 'x.x.x',
    },
  }),
  withReact: JSON.stringify({
    version,
    dependencies: {
      react: 'x.x.x',
    },
  }),
  withFlow: JSON.stringify({
    version,
    devDependencies: {
      flow: 'x.x.x',
    },
  }),
  withStencil: JSON.stringify({
    version,
    dependencies: {
      '@stencil/core': 'x.x.x',
    },
  }),
};

describe('init', () => {
  beforeEach(() => {
    // make sure terminal don't get noisy
    jest.spyOn(process.stdout, 'write').mockImplementation();
  });

  afterEach(() => {
    require('fs').__resetMockFiles();
    jest.clearAllMocks();
  });

  it('should guess angular projects', async () => {
    require('fs').__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withAngular);
    const targets = await guessTargets();
    expect(targets.Angular).toEqual(true);
  });

  it('should guess typescript projects', async () => {
    require('fs').__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withTypescript);
    const targets = await guessTargets();
    expect(targets.TypeScript).toEqual(true);
  });

  it('should guess react projects', async () => {
    require('fs').__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withReact);
    const targets = await guessTargets();
    expect(targets.React).toEqual(true);
  });

  it('should guess stencil projects', async () => {
    require('fs').__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withStencil);
    const targets = await guessTargets();
    expect(targets.Stencil).toEqual(true);
  });

  it('should guess flow projects', async () => {
    require('fs').__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withFlow);
    const targets = await guessTargets();
    expect(targets.Flow).toEqual(true);
  });

  it('should use angular related plugins when @angular/core is found', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withAngular);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    // silent
    jest.spyOn(console, 'log').mockImplementation();

    useInputs({
      onTarget: [ENTER], // confirm angular target
      onSchema: [ENTER], // use default
      onDocuments: [ENTER],
      onPlugins: [ENTER], // use selected packages
      onOutput: [ENTER], // use default output path
      onIntrospection: ['n', ENTER], // no introspection,
      onConfig: [ENTER], // use default config path
      onScript: ['graphql', ENTER], // use custom npm script
    });

    await init();

    expect(writeFileSpy).toHaveBeenCalledTimes(2);

    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;

    // should use default output path
    expect(config.generates['src/generated/graphql.ts']).toBeDefined();

    const output: any = config.generates['src/generated/graphql.ts'];
    expect(output.plugins).toContainEqual('typescript');
    expect(output.plugins).toContainEqual('typescript-operations');
    expect(output.plugins).toContainEqual('typescript-apollo-angular');
    expect(output.plugins).toHaveLength(3);

    // expected plugins
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-operations');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-apollo-angular');
    // should not have other plugins
    expect(Object.keys(pkg.devDependencies)).toHaveLength(4);
  });

  it('should use react related plugins when react is found', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withReact);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    // silent
    jest.spyOn(console, 'log').mockImplementation();

    useInputs({
      onTarget: [ENTER], // confirm react target
      onSchema: [ENTER], // use default
      onDocuments: [ENTER],
      onPlugins: [ENTER], // use selected packages
      onOutput: [ENTER], // use default output path
      onIntrospection: ['n', ENTER], // no introspection,
      onConfig: [ENTER], // use default config path
      onScript: ['graphql', ENTER], // use custom npm script
    });

    await init();

    expect(writeFileSpy).toHaveBeenCalledTimes(2);

    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;

    // should use default output path
    expect(config.generates['src/generated/graphql.tsx']).toBeDefined();

    const output: any = config.generates['src/generated/graphql.tsx'];
    expect(output.plugins).toContainEqual('typescript');
    expect(output.plugins).toContainEqual('typescript-operations');
    expect(output.plugins).toContainEqual('typescript-react-apollo');
    expect(output.plugins).toHaveLength(3);

    // expected plugins
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-operations');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-react-apollo');
    // should not have other plugins
    expect(Object.keys(pkg.devDependencies)).toHaveLength(4);
  });

  it('should use stencil related plugins when @stencil/core is found', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withStencil);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    // silent
    jest.spyOn(console, 'log').mockImplementation();

    useInputs({
      onTarget: [ENTER], // confirm angular target
      onSchema: [ENTER], // use default
      onDocuments: [ENTER],
      onPlugins: [ENTER], // use selected packages
      onOutput: [ENTER], // use default output path
      onIntrospection: ['n', ENTER], // no introspection,
      onConfig: [ENTER], // use default config path
      onScript: ['graphql', ENTER], // use custom npm script
    });

    await init();

    expect(writeFileSpy).toHaveBeenCalledTimes(2);

    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;

    // should use default output path
    expect(config.generates['src/generated/graphql.tsx']).toBeDefined();

    const output: any = config.generates['src/generated/graphql.tsx'];
    expect(output.plugins).toContainEqual('typescript');
    expect(output.plugins).toContainEqual('typescript-operations');
    expect(output.plugins).toContainEqual('typescript-stencil-apollo');
    expect(output.plugins).toHaveLength(3);

    // expected plugins
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-operations');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-stencil-apollo');
    // should not have other plugins
    expect(Object.keys(pkg.devDependencies)).toHaveLength(4);
  });

  it('should use typescript related plugins when typescript is found (node)', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withTypescript);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    // silent
    jest.spyOn(console, 'log').mockImplementation();

    useInputs({
      onTarget: [SELECT, ENTER], // confirm api target
      onSchema: [ENTER], // use default
      onPlugins: [ENTER], // use selected packages
      onOutput: [ENTER], // use default output path
      onIntrospection: ['n', ENTER], // no introspection,
      onConfig: [ENTER], // use default config path
      onScript: ['graphql', ENTER], // use custom npm script
    });

    await init();

    expect(writeFileSpy).toHaveBeenCalledTimes(2);

    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;

    // should use default output path
    expect(config.generates['src/generated/graphql.ts']).toBeDefined();

    const output: any = config.generates['src/generated/graphql.ts'];
    expect(output.plugins).toContainEqual('typescript');
    expect(output.plugins).toContainEqual('typescript-resolvers');
    expect(output.plugins).toHaveLength(2);

    // expected plugins
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-resolvers');
    // should not have other plugins
    expect(Object.keys(pkg.devDependencies)).toHaveLength(4); // 3 - because we have typescript package in devDeps
  });

  it('should have few default values', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withAngular);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    const logSpy = jest.spyOn(console, 'log').mockImplementation();
    const defaults = {
      schema: 'http://localhost:4000',
      documents: 'src/**/*.graphql',
      output: 'src/generated/graphql.ts',
      config: 'codegen.yml',
    };

    useInputs({
      onTarget: [ENTER], // confirm angular target
      onSchema: [ENTER], // use default
      onDocuments: [ENTER],
      onPlugins: [ENTER], // use selected packages
      onOutput: [ENTER], // use default output path
      onIntrospection: ['n', ENTER], // no introspection,
      onConfig: [ENTER], // use default config path
      onScript: ['graphql', ENTER], // use custom npm script
    });

    await init();

    const configFile = writeFileSpy.mock.calls[0][0] as string;
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;
    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);

    expect(pkg.scripts.graphql).toEqual(`graphql-codegen --config ${defaults.config}`);
    expect(configFile).toEqual(resolve(process.cwd(), defaults.config));
    expect(config.overwrite).toEqual(true);
    expect(config.schema).toEqual(defaults.schema);
    expect(config.documents).toEqual(defaults.documents);
    expect(config.generates[defaults.output]).toBeDefined();
    expect(logSpy.mock.calls[1][0]).toContain(`Config file generated at ${bold(defaults.config)}`);
  });

  it('should have few default values', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withAngular);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    const logSpy = jest.spyOn(console, 'log').mockImplementation();
    const options = {
      script: 'graphql',
      schema: './schema.ts',
      documents: 'graphql/**/*.graphql',
      output: 'graphql/index.ts',
      config: 'app-codegen.yml',
    };

    useInputs({
      onTarget: [ENTER], // confirm angular target
      onSchema: [options.schema, ENTER], // use default
      onDocuments: [options.documents, ENTER],
      onPlugins: [ENTER], // use selected packages
      onOutput: [options.output, ENTER], // use default output path
      onIntrospection: ['n', ENTER], // no introspection,
      onConfig: [options.config, ENTER], // use default config path
      onScript: [options.script, ENTER], // use custom npm script
    });

    await init();

    const configFile = writeFileSpy.mock.calls[0][0] as string;
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;
    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);

    expect(pkg.scripts[options.script]).toEqual(`graphql-codegen --config ${options.config}`);
    expect(configFile).toEqual(resolve(process.cwd(), options.config));
    expect(config.overwrite).toEqual(true);
    expect(config.schema).toEqual(options.schema);
    expect(config.documents).toEqual(options.documents);
    expect(config.generates[options.output]).toBeDefined();
    expect(logSpy.mock.calls[1][0]).toContain(`Config file generated at ${bold(options.config)}`);
  });

  it('custom setup', async () => {
    const fs = require('fs');
    fs.__setMockFiles(resolve(process.cwd(), 'package.json'), packageJson.withAngular);
    // make sure we don't write stuff
    const writeFileSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation();
    const logSpy = jest.spyOn(console, 'log').mockImplementation();
    const documents = 'graphql/*.ts';
    const script = 'generate:types';

    useInputs({
      onTarget: [ENTER], // confirm angular target
      onSchema: [ENTER], // use default
      onDocuments: [documents, ENTER],
      onPlugins: [ENTER], // use selected packages
      onOutput: [ENTER], // use default output path
      onIntrospection: ['y', ENTER], // no introspection,
      onConfig: [ENTER], // use default config path
      onScript: [script, ENTER], // use custom npm script
    });

    await init();

    expect(writeFileSpy).toHaveBeenCalledTimes(2);

    const pkg = JSON.parse(writeFileSpy.mock.calls[1][1] as string);
    const config = safeLoad(writeFileSpy.mock.calls[0][1] as string) as Record<string, any>;

    // config
    // should overwrite
    expect(config.overwrite).toEqual(true);
    // should match default schema
    expect(config.schema).toEqual('http://localhost:4000');
    // should match documents glob that we provided
    expect(config.documents).toEqual(documents);
    // should use default output path
    expect(config.generates['src/generated/graphql.ts']).toBeDefined();
    // should include introspection
    expect(config.generates['./graphql.schema.json']).toBeDefined();

    const output: any = config.generates['src/generated/graphql.ts'];
    expect(output.plugins).toContainEqual('typescript');
    expect(output.plugins).toContainEqual('typescript-operations');
    expect(output.plugins).toContainEqual('typescript-apollo-angular');

    // script name should match what we provided
    expect(pkg.scripts[script]).toEqual('graphql-codegen --config codegen.yml');
    // expected plugins
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-operations');
    expect(pkg.devDependencies).toHaveProperty('@graphql-codegen/typescript-apollo-angular');
    // should not have these plugins
    expect(pkg.devDependencies).not.toHaveProperty('@graphql-codegen/typescript-resolvers');

    // logs
    const welcomeMsg = logSpy.mock.calls[0][0];
    const doneMsg = logSpy.mock.calls[1][0];

    expect(welcomeMsg).toContain(`Welcome to ${bold('GraphQL Code Generator')}`);
    expect(doneMsg).toContain(`Config file generated at ${bold('codegen.yml')}`);
    expect(doneMsg).toContain(bold('$ npm install'));
    expect(doneMsg).toContain(bold(`$ npm run ${script}`));
  });

  describe('plugin choices', () => {
    function getAvailable(tags: Tags[]): string[] {
      return getPluginChoices({
        targets: tags,
      } as any).map((c: any) => c.value.value);
    }

    function getSelected(tags: Tags[]): string[] {
      return getPluginChoices({
        targets: tags,
      } as any)
        .filter((c: any) => c.checked)
        .map((c: any) => c.value.value);
    }

    function getPlugins(targets: Tags[]) {
      const tags: Tags[] = getApplicationTypeChoices({
        [Tags.angular]: targets.includes(Tags.angular),
        [Tags.react]: targets.includes(Tags.react),
        [Tags.stencil]: targets.includes(Tags.stencil),
        [Tags.browser]: targets.includes(Tags.browser),
        [Tags.node]: targets.includes(Tags.node),
        [Tags.typescript]: targets.includes(Tags.typescript),
        [Tags.flow]: targets.includes(Tags.flow),
        [Tags.vue]: targets.includes(Tags.vue),
      })
        .filter(c => c.checked)
        .reduce((all, choice) => all.concat(choice.value), []);

      return {
        available: getAvailable(tags),
        selected: getSelected(tags),
      };
    }

    it('node', () => {
      const { available, selected } = getPlugins([Tags.node]);

      // available
      expect(available).toHaveLength(6);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-resolvers');
      expect(available).toContainEqual('typescript-mongodb');
      expect(available).toContainEqual('typescript-document-nodes');
      expect(available).toContainEqual('flow');
      expect(available).toContainEqual('flow-resolvers');
      // selected
      expect(selected).toHaveLength(0);
    });

    it('node + typescript', () => {
      const { selected, available } = getPlugins([Tags.node, Tags.typescript]);

      // available
      expect(available).toHaveLength(4);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-resolvers');
      expect(available).toContainEqual('typescript-mongodb');
      expect(available).toContainEqual('typescript-document-nodes');
      // selected
      expect(selected).toHaveLength(2);
      expect(selected).toContainEqual('typescript');
      expect(selected).toContainEqual('typescript-resolvers');
    });

    it('node + flow', () => {
      const { selected, available } = getPlugins([Tags.node, Tags.flow]);

      // available
      expect(available).toHaveLength(2);
      expect(available).toContainEqual('flow');
      expect(available).toContainEqual('flow-resolvers');
      // selected
      expect(selected).toHaveLength(2);
      expect(selected).toContainEqual('flow');
      expect(selected).toContainEqual('flow-resolvers');
    });

    it('angular', () => {
      const { selected, available } = getPlugins([Tags.angular]);

      // available
      expect(available).toHaveLength(6);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-operations');
      expect(available).toContainEqual('typescript-apollo-angular');
      expect(available).toContainEqual('typescript-graphql-files-modules');
      expect(available).toContainEqual('typescript-document-nodes');
      expect(available).toContainEqual('fragment-matcher');
      // selected
      expect(selected).toHaveLength(3);
      expect(selected).toContainEqual('typescript');
      expect(selected).toContainEqual('typescript-operations');
      expect(selected).toContainEqual('typescript-apollo-angular');
    });

    it('react', () => {
      const { selected, available } = getPlugins([Tags.react]);

      // available
      expect(available).toHaveLength(8);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-operations');
      expect(available).toContainEqual('typescript-react-apollo');
      expect(available).toContainEqual('typescript-graphql-files-modules');
      expect(available).toContainEqual('flow');
      expect(available).toContainEqual('flow-operations');
      expect(available).toContainEqual('fragment-matcher');
      // selected
      expect(selected).toHaveLength(3);
      expect(selected).toContainEqual('typescript');
      expect(selected).toContainEqual('typescript-operations');
      expect(selected).toContainEqual('typescript-react-apollo');
    });

    it('react + typescript', () => {
      const { selected, available } = getPlugins([Tags.react, Tags.typescript]);

      // available
      expect(available).toHaveLength(6);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-operations');
      expect(available).toContainEqual('typescript-react-apollo');
      expect(available).toContainEqual('typescript-graphql-files-modules');
      expect(available).toContainEqual('typescript-document-nodes');
      expect(available).toContainEqual('fragment-matcher');
      // selected
      expect(selected).toHaveLength(3);
      expect(selected).toContainEqual('typescript');
      expect(selected).toContainEqual('typescript-operations');
      expect(selected).toContainEqual('typescript-react-apollo');
    });

    it('react + flow', () => {
      const { selected, available } = getPlugins([Tags.react, Tags.flow]);

      // available
      expect(available).toHaveLength(3);
      expect(available).toContainEqual('flow');
      expect(available).toContainEqual('flow-operations');
      expect(available).toContainEqual('fragment-matcher');
      // selected
      expect(selected).toHaveLength(2);
      expect(selected).toContainEqual('flow');
      expect(selected).toContainEqual('flow-operations');
    });

    it('stencil', () => {
      const { selected, available } = getPlugins([Tags.stencil]);

      // available
      expect(available).toHaveLength(6);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-operations');
      expect(available).toContainEqual('typescript-stencil-apollo');
      expect(available).toContainEqual('typescript-graphql-files-modules');
      expect(available).toContainEqual('typescript-document-nodes');
      expect(available).toContainEqual('fragment-matcher');
      // selected
      expect(selected).toHaveLength(3);
      expect(selected).toContainEqual('typescript');
      expect(selected).toContainEqual('typescript-operations');
      expect(selected).toContainEqual('typescript-stencil-apollo');
    });

    it('vanilla', () => {
      const { selected, available } = getPlugins([Tags.browser]);

      // available
      expect(available).toHaveLength(7);
      expect(available).toContainEqual('typescript');
      expect(available).toContainEqual('typescript-operations');
      expect(available).toContainEqual('typescript-graphql-files-modules');
      expect(available).toContainEqual('typescript-document-nodes');
      expect(available).toContainEqual('flow');
      expect(available).toContainEqual('flow-operations');
      expect(available).toContainEqual('fragment-matcher');
      // selected
      expect(selected).toHaveLength(0);
    });
  });

  describe('plugins', () => {
    it('should have correct plugin and package names', async () => {});
    plugins.forEach(pkg => {
      const { name } = require(`../../plugins/${pkg.pathInRepo}/package.json`);

      expect(pkg.package.replace('@graphql-codegen/', '')).toEqual(pkg.value);
      expect(pkg.package).toEqual(name);
    });
  });
});

function useInputs(inputs: {
  onTarget: string[];
  onSchema: string[];
  onDocuments?: string[];
  onPlugins: string[];
  onOutput: string[];
  onIntrospection: string[];
  onConfig: string[];
  onScript: string[];
}) {
  bddStdin(
    [].concat(
      inputs.onTarget,
      inputs.onSchema,
      inputs.onDocuments || [],
      inputs.onPlugins,
      inputs.onOutput,
      inputs.onIntrospection,
      inputs.onConfig,
      inputs.onScript
    )
  );
}
