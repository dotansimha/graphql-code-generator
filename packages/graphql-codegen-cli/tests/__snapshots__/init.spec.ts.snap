// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init custom setup 1`] = `
"
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  overwrite: true,
  schema: "http://localhost:4000",
  documents: "graphql/*.ts",
  generates: {
    "src/gql/": {
      preset: "client",
      plugins: []
    },
    "./graphql.schema.json": {
      plugins: ["introspection"]
    }
  }
};

export default config;
"
`;

exports[`init plugins suggestions for client-side setup should use angular related plugins when @angular/core is found 1`] = `
"
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  overwrite: true,
  schema: "http://localhost:4000",
  documents: "src/**/*.ts",
  generates: {
    "src/generated/graphql.ts": {
      plugins: ["typescript-apollo-angular"]
    }
  }
};

export default config;
"
`;

exports[`init plugins suggestions for client-side setup should use react related plugins when react is found 1`] = `
"
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  overwrite: true,
  schema: "http://localhost:4000",
  documents: "src/**/*.tsx",
  generates: {
    "src/gql/": {
      preset: "client",
      plugins: []
    }
  }
};

export default config;
"
`;

exports[`init plugins suggestions for client-side setup should use stencil related plugins when @stencil/core is found 1`] = `
"
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  overwrite: true,
  schema: "http://localhost:4000",
  documents: "src/**/*.graphql",
  generates: {
    "src/generated/graphql.tsx": {
      plugins: ["typescript", "typescript-operations", "typescript-stencil-apollo"]
    }
  }
};

export default config;
"
`;

exports[`init plugins suggestions non client-side setup should use typescript related plugins when typescript is found (node) 1`] = `
"
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  overwrite: true,
  schema: "http://localhost:4000",
  generates: {
    "src/generated/graphql.ts": {
      plugins: ["typescript", "typescript-resolvers"]
    }
  }
};

export default config;
"
`;

exports[`init should have few default values 1`] = `
"
import type { CodegenConfig } from '@graphql-codegen/cli';

const config: CodegenConfig = {
  overwrite: true,
  schema: "http://localhost:4000",
  documents: "src/**/*.tsx",
  generates: {
    "src/gql/": {
      preset: "client",
      plugins: []
    }
  }
};

export default config;
"
`;

exports[`init should have few default values 2`] = `
"overwrite: true
schema: "./schema.ts"
documents: "graphql/**/*.graphql"
generates:
  graphql/index.ts:
    preset: "client"
    plugins: []
  ./graphql.schema.json:
    plugins:
      - "introspection"
"
`;
